pupil_src/shared_modules/marker_tracker_3d/camera_localizer.py

```diff
@@ -0,0 +1,40 @@
+from marker_tracker_3d import math
+from marker_tracker_3d.localization import Localization
+
+
+class CameraLocalizer:
+    def __init__(self, storage):
+        self.storage = storage
+
+        self.localization = Localization(
+            self.storage.camera_model, self.storage.marker_model
+        )
+        self.min_number_of_markers_per_frame_for_loc = 2
+
+    def update(self):
+        self.update_camera_extrinsics()
+        self.update_camera_pose_matrix()
+        self.update_camera_trace()
+
+    def update_camera_extrinsics(self):
+        if len(self.storage.markers) >= self.min_number_of_markers_per_frame_for_loc:
+            self.storage.camera_extrinsics = self.localization.get_camera_extrinsics(
+                self.storage.markers,
+                self.storage.marker_extrinsics,
+                self.storage.camera_extrinsics_previous,
+            )
+        else:
+            self.storage.camera_extrinsics = None
+
+        if self.storage.camera_extrinsics is not None:
+            # Do not set camera_extrinsics_previous to None to ensure a decent initial guess for the next solve_pnp call
+            self.storage.camera_extrinsics_previous = self.storage.camera_extrinsics
+
+    def update_camera_pose_matrix(self):
```

Comment:
What was the reasoning for having the `camera_pose_matrix` included in `storage`? It is redundant with `camera_extrinsics` and could be computed on demand, right?

Reply:
`camera_pose_matrix` is mainly used for plotting camera frustum in the 3d window every frame, which is done in `UserInterface`. My though was, it is good not to have computation happening in `UserInterface`. So for now I just do `update_camera_pose_matrix` in `CameraLocalizer` and save it in the `storage`.
Same logic is applied to `camera_trace`